File: dsa_practice/invert_a_binary_tree.dart
class TreeNode {
  final int data;
  TreeNode? left;
  TreeNode? right;

  TreeNode({required this.data, this.left, this.right});

  @override
  String toString() {
    return 'TreeNode(data: $data, left:$left, right:$right)';
  }
}

 void invertTree(TreeNode? root) {
  if (root != null) {
    TreeNode? temp = root.left;
    root.left = root.right;
    root.right = temp;

    invertTree(root.left);
    invertTree(root.right);
  }
}

void main() {
  var tree =
      TreeNode(data: 2, left: TreeNode(data: 1), right: TreeNode(data: 3));
  invertTree(tree);
  print(tree);
}

File: dsa_practice/invert_a_binary_tree.py
def invertTree(root):
    """
    :type root: TreeNode
    :rtype: TreeNode
    """
    if root:
        temp = root.left
        root.left = root.right
        root.right = temp

        # recursively swap it's children if it has
        invertTree(root.left)
        invertTree(root.right)
    return root  

# time complexity: O(n), we will visit every Node
# space complexity: O(log n)[best case; balanced tree] or O(n)[worst case; inbalanced tree] 

